### PR TITLE
feat: start implementation of csvw builder

### PR DESCRIPTION
**Pull request by tpluscode**

### CODE REVIEW

File: ui/src/api/mixins/ApiResource.ts [hunk 1]
```diff
@@ -32,6 +32,10 @@ export function apiResourceMixin (rootUrl: string, separator: string): Mixin {
       get actions (): Actions {
         return commonActions(this, this._additionalActions)
       }
+
+      get inlineRepresentation () {
```
Comment (martinmaillard) [inline ui/src/api/mixins/ApiResource.ts:36]: Not sure I understand this

**Comment by martinmaillard** [comment]
If I put a space in the `targetProperty` of a column mapping, it crashes the pipeline. Not sure if we should prevent that or just escape spaces.

File: ui/src/api/mixins/ApiResource.ts (hunk 1 above)
Comment (tpluscode) [reply ui/src/api/mixins/ApiResource.ts:36]: Right, this is a little annoying technicality, for which I could not find a good solution. 

Take [Table#columnMappings](https://github.com/zazuko/cube-creator/pull/169/files#diff-952eebae7242509177c8aa88b60ec294a4c0fd9f04e5a20554993a7f2464e184R51) for an example. It is typed as an array of `Link<ColumnMapping>`. A link is deliberately typed as a plain resource with a `load` method to prevent accidentally adding triples about the linked resource inside the Table's graph.

But reconsidering this, an explicit cast can work just as well without adding meaningless properties

Comment (martinmaillard) [reply ui/src/api/mixins/ApiResource.ts]: TIL what a `Link` is :)

File: fuseki/sample-ubd.trig [hunk 2]
```diff
@@ -216,7 +216,7 @@ graph <cube-project/ubd/csv-source/ubd> {
 
   <cube-project/ubd/csv-source/ubd/column/year>
     a csvw:Column ;
-    schema:name "YEAR" ;
+    schema:name "year" ;
```
Comment (martinmaillard) [inline fuseki/sample-ubd.trig:219]: Why this change?

File: fuseki/sample-ubd.trig [hunk 3]
```diff
@@ -386,8 +386,7 @@ graph <cube-project/ubd/csv-mapping/table-station> {
     cc:csvw <cube-project/ubd/csv-mapping/table-station/csvw> ;
     schema:name "Station" ;
     schema:color "#BBBBBB" ;
-    # cc:identifierTemplate "station/{STATION_ID}" ;
-    cc:identifierTemplate "{station_id}-{year}-{aggregation_id}" ;
+    cc:identifierTemplate "station/{STATION_ID}" ;
```
Comment (martinmaillard) [inline fuseki/sample-ubd.trig:389]: The `STATION_ID` variable is wrong now that you downcased the column names

File: packages/model/Table.ts [hunk 4]
```diff
@@ -18,7 +19,7 @@ export interface Table extends RdfResource {
   csvMapping: Link<CsvMapping>
   identifierTemplate: string
   color: string
-  columnMappings: ColumnMapping[]
+  columnMappings: Array<Link<ColumnMapping>>
```
Comment (martinmaillard) [inline packages/model/Table.ts:22]: In the UI, they are not really `Link`s right? I see that you fixed it by "convincing" typescript that they are full resources but I wonder if we are not already hitting some of the limits of sharing these models.

File: apis/core/lib/csvw-builder/index.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+      warning(`Column ${columnMapping.sourceColumn.id} not found`)
+      continue
+    }
+
+    yield mappedColumn({ csvMapping, columnMapping, column, source, resources })
+  }
+
+  for (const [, column] of unmappedColumns) {
+    yield unmappedColumn({ column })
+  }
```
Comment (lucafurrer) [inline apis/core/lib/csvw-builder/index.ts:68]: I find it rather strange that the function `buildColumns` returns the ignored columns.

File: apis/core/lib/csvw-builder/index.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  const source = await resources.getResource<CsvSource>(table.csvSource?.id)
+  if (!source) {
+    throw new Error(`Source not found for table ${table.id.value}`)
+  }
+  const csvMapping = await resources.getResource<CsvMapping>(table.csvMapping.id)
+  if (!csvMapping) {
+    throw new Error(`CSV Mapping resource not found for table ${table.id.value}`)
+  }
+
+  const column: Initializer<Csvw.Column>[] = [{
```
Comment (lucafurrer) [inline apis/core/lib/csvw-builder/index.ts:81]: since it is an array column**s**?

File: packages/model/Table.ts (hunk 4 above)
Comment (tpluscode) [reply packages/model/Table.ts:22]: This is just annotation so yes, the properties are actual resource objects.

That aside, I have been thinking about this and it is true that this is a contention point between the two usages of a shared model. In the long run I would rather that the UI model interface are for example generated from a SHACL shapes.

This would especially useful for the designer view.

File: apis/core/lib/csvw-builder/index.ts (hunk 6 above)
Comment (lucafurrer) [reply apis/core/lib/csvw-builder/index.ts:81]: Does column not contain several columns? 
`Initializer<Csvw.Column>[]` and  `column.push(csvwColumn)` made me think it is an array of columns

File: packages/model/Table.ts (hunk 4 above)
Comment (martinmaillard) [reply packages/model/Table.ts:22]: Yeah... let's not go into this for now ;)

File: fuseki/sample-ubd.trig (hunk 3 above)
Comment (tpluscode) [reply fuseki/sample-ubd.trig:389]: Done. Although we might want to sync the test dataset and the sample file

**Review by martinmaillard — APPROVED** [review]
